### PR TITLE
PERF: leverage tzlocal package to provide 2000x speedup for dateutil.tz.tzlocal operations

### PR DESCRIPTION
**Pull request by qwhelan**
This is a follow-up to https://github.com/pandas-dev/pandas/pull/24491, where it was noted that `tzlocal` operations are notably slower:
```
[ 53.12%] ··· timeseries.DatetimeIndex.time_add_timedelta                                                                                                                 ok
[ 53.12%] ··· ============ =============
               index_type
              ------------ -------------
                  dst         471±8μs
                repeated    3.52±0.05ms
                tz_aware    3.78±0.07ms
              **tz_local     5.77±0.01s**
                tz_naive    1.98±0.03ms
              ============ =============
```
This PR makes them indistinguishable from other timezones:
```
[ 60.00%] ··· timeseries.DatetimeIndex.time_add_timedelta                                                                                                                 ok
[ 60.00%] ··· ============ =============
               index_type
              ------------ -------------
                  dst         465±8μs
                repeated    3.55±0.07ms
                tz_aware    3.65±0.05ms
                tz_local    3.68±0.04ms
                tz_naive    1.81±0.05ms
              ============ =============
```
To do this, we add a new optional dependency: `tzlocal`. This package queries the OS to try and identify a `pytz` compatible timezone name for the local timezone. If this is successful, we can use the `pytz` timezone instead of `dateutil.tz.tzlocal()`, yielding a huge speedup as the latter invokes `datetime.datetime()`. In cases where the local timezone is also UTC, further speedups are possible.

In all cases, the `tzlocal()` object is maintained as the `.tz` to preserve any semantic meaning.

One unresolved issue is an incompatibility with `python-dateutil` versions prior to `2.6.0`. I'm unable to reproduce the error locally, but can repeatably get this error on Travis:
```
    def test_timetz_accessor(self, tz_naive_fixture):
        # GH21358
        tz = timezones.maybe_get_tz(tz_naive_fixture)
    
        expected = np.array([time(10, 20, 30, tzinfo=tz), pd.NaT])
    
        index = DatetimeIndex(['2018-06-04 10:20:30', pd.NaT], tz=tz)
        result = index.timetz
    
>       tm.assert_numpy_array_equal(result, expected)
pandas/tests/indexes/datetimes/test_timezones.py:824: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/core/dtypes/missing.py:404: in array_equivalent
    ensure_object(left.ravel()), ensure_object(right.ravel()))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   if not (PyObject_RichCompareBool(x, y, Py_EQ) or
E   TypeError: can't compare offset-naive and offset-aware times
```
While this only shows up on the 2.7 build, I believe this is due to it being the only one actually building against `python-dateutil=2.5.0`. Any version `>=2.6.0` does not exhibit this error.

Finally, here is the `asv` benchmark against `v0.24.0rc1`:
```
$ asv compare v0.24.0rc1 HEAD -s --sort ratio --only-changed
       before           after         ratio
     [fdc4db25]       [56e5cc0d]
     <v0.24.0rc1^0>       <tzlocal>
-        68.4±5μs         17.9±1μs     0.26  timeseries.DatetimeIndex.time_get('tz_local')
-      5.54±0.02s          188±2ms     0.03  timeseries.DatetimeIndex.time_to_pydatetime('tz_local')
-      7.13±0.02s       30.2±0.7ms     0.00  timeseries.DatetimeIndex.time_normalize('tz_local')
-      6.00±0.05s       23.7±0.3ms     0.00  timeseries.DatetimeIndex.time_to_time('tz_local')
-      6.00±0.01s       21.3±0.3ms     0.00  timeseries.DatetimeIndex.time_to_date('tz_local')
-      5.98±0.06s      3.68±0.04ms     0.00  timeseries.DatetimeIndex.time_add_timedelta('tz_local') 
-      5.96±0.05s      2.49±0.05ms     0.00  timeseries.DatetimeIndex.time_timeseries_is_month_start('tz_local')
```

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by mroeschke — CHANGES_REQUESTED** [review]
Overall we need to consider whether this is worth bumping up our min dateutil version and adding an optional dependency to accelerate only `dateutil.tz.tzlocal`

Alternatively, could `tzlocal` be vendored?

File: pandas/_libs/tslibs/timezones.pyx [hunk 1]
```diff
@@ -34,6 +36,38 @@ cdef inline bint is_tzlocal(object tz):
...
+cpdef object get_tzlocal_tz(object tz):
+    if have_tzlocal_package():
+        global tzlocal_tz
+        if tzlocal_tz is None:
+            local_tz = tzlocal_package.get_localzone()
+            tzlocal_tz = local_tz.zone
+
+        try:
+            return pytz.timezone(tzlocal_tz)
+        except pytz.exceptions.UnknownTimeZoneError:
```
[the remark this answers is not in the document]
Comment (qwhelan) [reply pandas/_libs/tslibs/timezones.pyx]: `tzlocal` will report the system tz even if it isn't a valid `pytz` tz. For example, we'll hit this case if `/etc/timezone` contains `foo/bar` or, more likely, an ambiguous user timezone like `STD`.

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 2]
```diff
@@ -820,7 +820,8 @@ def test_timetz_accessor(self, tz_naive_fixture):
 
         index = DatetimeIndex(['2018-06-04 10:20:30', pd.NaT], tz=tz)
         result = index.timetz
-
+        print('result', result)
```
[the remark this answers is not in the document]
Comment (qwhelan) [reply pandas/tests/indexes/datetimes/test_timezones.py]: Yup, will clean up

File: pandas/_libs/tslibs/timezones.pyx [hunk 3]
```diff
@@ -34,6 +36,38 @@ cdef inline bint is_tzlocal(object tz):
     return isinstance(tz, _dateutil_tzlocal)
 
 
+cpdef bint have_tzlocal_package():
```
[the remark this answers is not in the document]
Comment (qwhelan) [reply pandas/_libs/tslibs/timezones.pyx]: Absolutely

**Review by jreback — CHANGES_REQUESTED** [review]
we just moved min version for python dateutil
how hard is it to vendor tzlocal?

**Comment by qwhelan** [comment]
@mroeschke Agreed, it's worth mentioning that this mostly fixes the top performance regression between `v0.23.4` and `v0.24.0rc1`:
```
       before           after         ratio
     [04095216]       [fdc4db25]
     <v0.23.4^0>       <v0.24.0rc1^0>
+     3.91±0.01ms       18.8±0.01s  4819.04  timeseries.DatetimeIndex.time_add_timedelta('tz_local')
```
It'd be relatively simple to vendor it as it's only a handful of files. It's MIT licensed, so should be fine but worth mentioning now.

I haven't been able to reproduce the `python-dateutil` conflict locally thus far, but I'll dig into that further to see if there's an easy workaround.

**Comment by jreback** [comment]
if we bump to 2.6.0 would u still need tzlocal?

**Comment by qwhelan** [comment]
@jreback Yes, there's just a hard-to-reproduce conflict with 2.5.0-2.5.3. Not sure how serious it is at this point.

@jbrockmendel Regarding accuracy, it's pulling from system files like `/etc/timezone`, `/var/db/zoneinfo`, etc so should accurately reflect the system timezone. However, given that it's checking so many places it's entirely possible a user has different timezones configured in different files.

There's no distinction between `America/Los_Angeles` and `US/Pacific` other than the latter being deprecated (https://en.wikipedia.org/wiki/List_of_tz_database_time_zones).

**Comment by mroeschke** [comment]
Although tzlocal hurts portability (unless that's the intension), I think pandas should ensure that timezones in == timezones out and shouldn't make any assumptions or transformation when returning the timezone (i.e tzlocal in, tzlocal out; US/Pacific in, US/Pacific out), which I can see your PR does @qwhelan .

Overall agreed that I'd be great to address this regression. Shot in the dark: I think the proper way to compare dateutil tzs is with `is` and not `==`, so maybe that's why you're seeing that error?

**Comment by qwhelan** [comment]
@jreback I'm finally able to reproduce `dateutil` 2.5.0 issue locally and think I have it resolved without needing to bump the dependency. Where would be the appropriate directory to vendor `tzlocal`?

**Comment by jreback** [comment]
put in pandas/tseries/_tzlocal.py i think - see how that works

**Comment by mroeschke** [comment]
@jreback How about `pandas/_libs/tslibs/*`? Could give way to cythonize some things?

**Comment by jreback** [comment]
sure that’s ok too

**Comment by jbrockmendel** [comment]
> Could give way to cythonize some things?

Is there anything in particular that bears cythonizing?

**Comment by mroeschke** [comment]
Haven't looked closely, but I imagine most all functions here can be `cdef`'d given that it just interacts in `timezones.pyx`?

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -17,7 +17,7 @@ cimport numpy as cnp
 from numpy cimport int64_t
 cnp.import_array()
 
-
+cimport pandas._libs.tslibs
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/offsets.pyx]: Why?

File: pandas/_libs/src/tzlocal/__init__.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import sys
```
Comment (jreback) [inline pandas/_libs/src/tzlocal/__init__.py]: I wouldn't put this in src (which is not a real package, nor do we want it to be), rather just in ``pandas/_libs/tzlocal`` is fine

Comment (jreback) [inline pandas/_libs/src/tzlocal/__init__.py]: can you make it clear that this is a vendored copy

File: pandas/_libs/tslibs/timezones.pyx [hunk 6]
```diff
@@ -34,6 +37,25 @@ cdef inline bint is_tzlocal(object tz):
     return isinstance(tz, _dateutil_tzlocal)
 
 
+cpdef object get_tzlocal_tz(object tz):
+    global tzlocal_tz
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx]: is this to prevent import issues?

**Review by jreback — CHANGES_REQUESTED** [review]
so I wouldn't change anything in the vendored copy (its hard to tell if you did); I just saw a comment about 'cythonizing it'. Just drop it in place.

File: pandas/_libs/tslibs/offsets.pyx (hunk 4 above)
Comment (qwhelan) [reply pandas/_libs/tslibs/offsets.pyx]: Cython was making noise about this when running under 2.7; looks like it doesn't raise the same error under 3. Will remove this.

File: pandas/_libs/tslibs/timezones.pxd [hunk 7]
```diff
@@ -3,6 +3,9 @@
 cpdef bint is_utc(object tz)
 cdef bint is_tzlocal(object tz)
 
+cpdef object get_tzlocal_tz(object tz)
+cpdef _set_tzlocal_tz(object tz)
```
[the remark this answers is not in the document]
Comment (qwhelan) [reply pandas/_libs/tslibs/timezones.pxd]: It's currently being used to assist in testing, where we cycle through all `pytz` timezones and check equivalence against `tzlocal`.

**Comment by WillAyd** [comment]
@qwhelan can you merge master and address build failures?

File: pandas/_libs/src/tzlocal/__init__.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/src/tzlocal/__init__.py]: alternative would be something like `pandas/vendored/`.  `pd.io.clipboard` would also belong in such a directory.

Comment (jreback) [reply pandas/_libs/src/tzlocal/__init__.py]: call this ``pandas/_vendored/`` (explicity private)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master and update

**Comment by jreback** [comment]
can you merge master and update

**Comment by jreback** [comment]
closing as stale. nice idea though, pls ping if you can continue.

**Pull request closed without merging**